web/pdf_page_view.js

```diff
@@ -569,17 +580,35 @@ class PDFPageView {
       }
     }
 
+    const postponeDrawing = drawingDelay >= 0 && drawingDelay < 1000;
+
     if (this.canvas) {
       if (
+        postponeDrawing ||
         this.useOnlyCssZoom ||
         (this.hasRestrictedScaling && isScalingRestricted)
       ) {
+        if (
+          postponeDrawing &&
+          this.renderingState === RenderingStates.RUNNING
```

Comment:
Is this check actually sufficient, since there's also the `PAUSED` state (where it also makes sense to cancel rendering)?
Do we perhaps just want to do this whenever `this.renderingState !== RenderingStates.FINISHED` instead? Obviously that'd include the `INITIAL` state, but given that no rendering happens in that case rendering-cancellation becomes a no-op then.

Reply:
I managed to have an exception (not easy to reproduce: I reload and zoom up/down before loading has finished) because we're cancelling a paused task and then at some point `resume` was called but it has been nulled in `cancelRendering`:
```
Uncaught (in promise) TypeError: view.resume is not a function
    renderView http://localhost:8888/web/pdf_rendering_queue.js:189
    forceRendering http://localhost:8888/web/pdf_viewer.js:1672
    promise callback*forceRendering http://localhost:8888/web/pdf_viewer.js:1671
    renderHighestPriority http://localhost:8888/web/pdf_rendering_queue.js:80
    update http://localhost:8888/web/pdf_viewer.js:1464
    _scrollUpdate http://localhost:8888/web/pdf_viewer.js:1039
    viewAreaElementScrolled http://localhost:8888/web/ui_utils.js:178
    debounceScroll http://localhost:8888/web/ui_utils.js:163
    watchScroll http://localhost:8888/web/ui_utils.js:191
    PDFViewer http://localhost:8888/web/pdf_viewer.js:310
    _initializeViewerComponents http://localhost:8888/web/app.js:482
    initialize http://localhost:8888/web/app.js:239
    run http://localhost:8888/web/app.js:643
    webViewerLoad http://localhost:8888/web/viewer.js:222
    promise callback*webViewerLoad http://localhost:8888/web/viewer.js:221
    <anonymous> http://localhost:8888/web/viewer.js:258
```
And if I add `this.renderingState = RenderingStates.INITIAL;` after the `cancelRendering` like we do in `reset`, sometimes the space between two pages disappears during zooming. I think it'd make sense to reset the state.
I'll have a look on that stuff tomorrow, but if you've any ideas I take.

Reply:
With the caveat that I've gotten sick again and possible isn't thinking straight; isn't there potentially a "bigger picture" issue here?

If the page isn't `FINISHED` rendering when we get to here, won't the next call to `PDFViewer.update` (which happens e.g. via the "scalechanging" event listener) actually force it to start re-rendering *immediately* rather than being delayed?
Hence it seems to me that we'd need a way to both cancel any pending rendering here *and* also ensure that it won't attempt to start re-rendering before delay actually runs out.

Reply:
Actually, do we perhaps just want to set `this.renderingState = RenderingStates.FINISHED;` regardless here to avoid this?

Reply:
I was adding a new state `POSTPONED` but I think you're right and using `FINISHED` is good enough.